### PR TITLE
Migration of Dataset details mako to VueJS

### PR DESCRIPTION
**Pull request by anuprulez**
This PR migrates the mako of Dataset details to VueJS. It follows https://github.com/galaxyproject/galaxy/pull/4978 to create a .vue file (single file component) to replace the corresponding mako. Also, for creating a section of "Tool Parameters", one recursive method (which creates a part of HTML for this section) has been moved from the mako to its respective controller so as to avoid refactoring the server side code in the .vue component file.

Your suggestions are welcome!

Thanks @dannon and @bgruening 

Edit: The markup has one part which is added dynamically and there is a need to bind its click event to
a method defined in the vue component. Due to the dymanic nature of this part (of HTML), it needs to be compiled again so that it knows its event. I tried using `Vue.compile` to compile the dynamic part of the whole HTML and render again but I could not succeed. To solve it, I have used "Updated" method (which calls once the DOM is rendered) and inside it I have included a jQuery click event which replaces the view's model to have the details of the new dataset. There is another way (which should work as well but not verified) to achieve it through `$nextTick`

### CODE REVIEW

File: client/galaxy/scripts/components/show_params.vue [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        </table>
+       
+        <span v-if="hasParameterErrors"> {{ showError( `One or more of your original parameters may no longer be valid or displayed properly.` ) }} </span>
+
+        <h3>Inheritance Chain</h3>
+        <div class="inherit inherit-chain"> {{ hdaData.name }} </div>
+        <div v-for="dep in inheritChain">
+            <div class="inherit-chain-item">&uarr;</div>
+            <div class="inherit">
+                {{ dep[0].name }} in {{ dep[1] } <br/>
```
Comment (guerler) [inline client/galaxy/scripts/components/show_params.vue]: Closing bracket is missing.

File: client/galaxy/scripts/components/show_params.vue [hunk 2]
```diff
@@ -0,0 +1,209 @@
+<template>
+    <div id="show-dataset-params" class="dataset-metadata" v-cloak>
+        <h2>
+            {{ toolName }}
+        </h2>
+        <h3>Dataset Information</h3>
+        <table class="tabletip" id="dataset-details">
+            <tbody>
+                <tr><td> Number:</td><td> {{ hdaData.number }} </td></tr>
+                <tr><td> Name:</td><td> {{ hdaData.name }} </td></tr>
```
Comment (jmchilton) [inline client/galaxy/scripts/components/show_params.vue]: For other people new to Vue and reviewing this - I Googled it and Vue is automatically escaping things - very nice!

File: client/galaxy/scripts/components/show_params.vue (hunk 1 above)
Comment (anuprulez) [reply client/galaxy/scripts/components/show_params.vue]: Fixed. Thanks!

**Comment by jmchilton** [comment]
Pushed a fix to a couple Selenium tests based on the changes in this branch. If it doesn't come back all green I can take another crack at them as well.

**Comment by jmchilton** [comment]
So the history view test failure seems to be legitimate and fixable by tweaking Selenium. Below is the browser console logging picked up by the test:

```
[
 {
  "source": "network", 
  "message": "http://localhost:8295/static/scripts/components/show_params.vue.js?v=1511556040 - Failed to load resource: the server responded with a status of 404 (Not Found)", 
  "timestamp": 1511557898647, 
  "level": "SEVERE"
 }, 
 {
  "source": "javascript", 
  "message": "http://localhost:8295/static/scripts/libs/require.js?v=1511556017 0:1066 Uncaught Error: Script error for \"components/show_params.vue\", nee\u2026http://requirejs.org/docs/errors.html#scripterror", 
  "timestamp": 1511557898660, 
  "level": "SEVERE"
 }
]
```

I guess this is like citations not working in workflows until it was bundled. This can be replicated in browser by publishing a history and then going to "Shared Data -> Histories" and clicking on it.

Unrelated but another I thing I noticed in testing is that show_params is still used in the history structure view and doesn't work with this PR. To try this out click on the history options button and click "Show Structure".

**Comment by anuprulez** [comment]
@jmchilton Thanks a lot for your inputs. I looked for "show_params" in the repository and found other occurrences of this mako like in `templates/webapps/galaxy/data_manager/view_job.mako, templates/webapps/galaxy/history/display_structured.mako...`

I will have a look at these.

File: client/galaxy/scripts/mvc/dataset/dataset-li.js [hunk 3]
```diff
@@ -252,21 +254,30 @@ var DatasetListItemView = _super.extend(
...
+                    if ( Galaxy.frame && Galaxy.frame.active ) {
                         Galaxy.frame.add({
-                            title: _l("Dataset details"),
-                            url: this.href
+                            title: _l("Dataset details")
                         });
-                        ev.preventDefault();
-                        ev.stopPropagation();
+                        let $elFrame = $( ".corner.frame" ),
+                            $elLatestFrame = $( $elFrame[ $elFrame.length - 1 ] ).find( ".f-content" );
```
Comment (guerler) [inline client/galaxy/scripts/mvc/dataset/dataset-li.js]: We should not rely on internal css classes of the scratchbook viewer.

File: lib/galaxy/webapps/galaxy/controllers/dataset.py [hunk 4]
```diff
@@ -1081,15 +1084,151 @@ def show_params(self, trans, dataset_id=None, from_noframe=None, **kwd):
...
+            plugin_metric_displays = [trans.app.job_metrics.format(metric.plugin, metric.metric_name, metric.metric_value) for metric in plugin_metrics]
+            plugin_metric_displays = sorted(plugin_metric_displays, key=lambda pair: pair[0])
+            job_metrics["plugin_metric_displays"][plugin] = plugin_metric_displays
+
+        # create template for tool parameters using a recursive method
+        tool_parameter_template = ''
+        if params_objects and tool:
+            tool_parameter_template = self.inputs_recursive(trans, tool.inputs, params_objects, depth=1, upgrade_messages=upgrade_messages)
+        elif params_objects is None:
+            tool_parameter_template = '<tr><td colspan="3">Unable to load parameters.</td></tr>'
```
Comment (guerler) [inline lib/galaxy/webapps/galaxy/controllers/dataset.py]: Can we parse the data as dictionary and then build the html on the client side?

File: client/galaxy/scripts/mvc/dataset/dataset-li.js (hunk 3 above)
Comment (anuprulez) [reply client/galaxy/scripts/mvc/dataset/dataset-li.js]: I will look into it... thanks!

File: lib/galaxy/webapps/galaxy/controllers/dataset.py (hunk 4 above)
Comment (anuprulez) [reply lib/galaxy/webapps/galaxy/controllers/dataset.py]: I thought about it but this piece of HTML generation recursively has a lot of clauses. Maybe for each line of HTML one clause/condition based on which the entire chunk of HTML is generated. 
For now, I have created a util file which takes care of HTML template generation at the server. Once I am sure all the relevant makos have been migrated, I will come back to this.

**Comment by anuprulez** [comment]
After migrating two makos, to move ahead in this PR, I need to migrate a few more related to the data managers at `webapps/galaxy/data_managers/`
- index

dmindex

- manage_data_manager

viewjob

- view_job

dataset

The last one contains a link to the "view details" of HDA (as "view complete info" link). Since migrating "view details" (show_params.mako) is the main work of this PR, it is important to tackle all the makos up the tree which are connected involving all these aforementioned admin makos. 

@jmchilton @dannon @bgruening 

I can migrate these makos. Also I saw a "vue" for data manager table. May be I can use it. 
What are your views?

Thank you!

**Comment by martenson** [comment]
This is marked as WIP so I am pushing it to 18.05.

**Comment by anuprulez** [comment]
Through this pull request, I would like to share two situations I came across (which I have not been able to solve):

- While converting any mako to vueJS view, we are making an asynchronous call through a vue component to fetch required JSON to build the view. This works fine and the views are created. But, the test cases, it seems, do not have this step (which is why tool shed and selenium tests are failing for this pull request). For example, when a "manage data manager" link is clicked (with ref to this PR) using a client side URL, it fetches a skeleton of HTML (parent view in which our tables related to data managers are not yet created). Through this view, we fetch the json related to data managers to create the child view. In the test cases, we try to find the HTML elements of child view (e.g. table#dataset-details) in the parent view (accessed using Twill) which are not yet present. That's why the some of the toolshed (related to data managers - test_0030_verify_data_manager_tool) and selenium tests are failing. Previously with mako files, the parent view contained the complete HTML.
I tried solving this using selenium webdriver with implicit waits and time.sleep but they did not help.
We should be able to fetch the complate html rendered using a url (including after all the asynchronous calls).

- This PR works to remove the mako files related to datasets. For mako files, there are separate URLs (datasets/{id}/show_params) for viewing the details ("view details"). These makos are opened in iframes using these urls. But now, with vueJS we do it asynchronously and thereby removing these urls. I thought of keeping these urls for "view details" of each dataset but when these are accessed, the Galaxy reloads (the client side urls). Working with separate urls which reload the application can cause a problem when a user tries to view the metadata (view details) in the scratchbook. As soon as the page reloads (to pull metadata for a dataset), scratchbook is reset. I think it is great to have separate url for each dataset to view its metadata and we should be able to open this view in the scratchbook at the same time.

Please share if you have any insights. Thank you!

ping @guerler @bgruening @jmchilton @martenson @dannon

**Comment by dannon** [comment]
@anuprulez Thanks for the work on this!  I'll check the tests out first thing tomorrow and see what I can come up with.

@anuprulez Which reload is triggering a full page reload?  Is it just not scoped correctly (That is, targeting the parent instead of the iframe?)?  I can look into this more tomorrow, too, if you aren't sure.

**Comment by anuprulez** [comment]
@dannon Thank you!
This PR has become a bit bigger and complicated to tackle at once. So, I want to break it down into smaller problems and will see how they go.

**Comment by jmchilton** [comment]
I'd just remove the tool shed test - Twill is never going to be able to handle this async stuff well enough I don't think.

For the Selenium things - I'd avoid the sleeps and such. Just wait for the condition you are hoping for. 

For instance, wait for some selector that isn't there until the page is fully rendered:

```python
        title = self.wait_for_selector(".page-body h3")
```

Or wrap the whole condition in a retry as long as it is a check that doesn't mutate the page:

```python
        @retry_assertion_during_transitions
        def check_name():
            name = self.workflow_index_name()
            assert 'CoolNewName' == name, name
```

It looks like you are waiting on an element that never shows up in the failing test - are you sure it should still be there?

I'd really like to add an entry point to show job information for output collections - I was going to work on it and then I remembered this PR. I really need to use some stuff from here - anything I can do to speed this along?

**Comment by anuprulez** [comment]
@jmchilton Thanks a lot for your comments. Yes, the tests (selenium ones) wait for particular HTML structures and they do not get them (because we use client-side url and catch them in `analysis.js`, create a view using `vuejs` and then make a silent async call to fetch data in order to create the complete HTML structure for a feature. Here at this stage, we need tests to ensure the correctness of the created HTML).

This PR became too complicated to tackle while migrating 4-5 makos at once mixing the history stuff with data managers. So, in a follow-up PR #5317, I partly replicated it to migrate two makos related to history. 

I would recommend to close this PR and tackle the challenges posed by it separately. If anything is useful here for other tasks, it is perfectly fine to be taken forward. 

Thank you :)

**Comment by dannon** [comment]
@anuprulez Sounds good.  I'll close this out and we can use parts of it in smaller efforts.

**Pull request closed without merging**